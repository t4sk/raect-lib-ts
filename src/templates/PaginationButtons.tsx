import React from "react";
import { Button, Label } from "semantic-ui-react";
import styles from "./PaginationButtons.module.css";

interface Props {
  count: number;
  skip: number;
  limit: number;
  onChangePage: (params: { page: number; skip: number }) => any;
}

export const PaginationButtons: React.FC<Props> = (props) => {
  const { count, skip, limit, onChangePage } = props;

  const current = Math.floor(skip / limit) + 1;
  const total = Math.max(Math.ceil(count / limit), 1);

  function onClick(page: number) {
    onChangePage({
      page,
      skip: (page - 1) * limit,
    });
  }

  return (
    <div className={styles.component}>
      <Button
        size="mini"
        basic
        disabled={current === 1}
        onClick={() => onClick(1)}
      >
        First
      </Button>

      <Button
        data-testid="prev"
        size="mini"
        basic
        disabled={current === 1}
        icon="chevron left"
        onClick={() => onClick(current - 1)}
      />

      <Label basic>
        <div>
          {(skip + 1).toLocaleString()}-
          {Math.min(skip + limit, count).toLocaleString()} of{" "}
          {count.toLocaleString()}
        </div>
      </Label>

      <Button
        data-testid="next"
        size="mini"
        basic
        disabled={current === total}
        icon="chevron right"
        onClick={() => onClick(current + 1)}
      />
      <Button
        size="mini"
        basic
        disabled={current === total}
        onClick={() => onClick(total)}
      >
        Last
      </Button>
    </div>
  );
};

export default PaginationButtons;
