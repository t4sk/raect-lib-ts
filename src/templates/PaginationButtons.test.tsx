import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { PaginationButtons } from "./PaginationButtons";

const props = {
  count: 100,
  skip: 40,
  limit: 20,
  onChangePage: jest.fn(),
};

beforeEach(() => {
  props.onChangePage.mockClear();
});

test("it renders", () => {
  const { container } = render(<PaginationButtons {...props} />);

  expect(container).toMatchSnapshot();
});

test("render first page", () => {
  const { container } = render(<PaginationButtons {...props} skip={0} />);

  expect(container).toMatchSnapshot();
});

test("render last page", () => {
  const { container } = render(<PaginationButtons {...props} skip={80} />);

  expect(container).toMatchSnapshot();
});

test("on click first", () => {
  const { getByText } = render(<PaginationButtons {...props} />);

  fireEvent.click(getByText("First"));

  expect(props.onChangePage.mock.calls.length).toEqual(1);
  expect(props.onChangePage.mock.calls[0][0]).toEqual({
    page: 1,
    skip: 0,
  });
});

test("on click prev", () => {
  const { getByTestId } = render(<PaginationButtons {...props} />);

  fireEvent.click(getByTestId("prev"));

  expect(props.onChangePage.mock.calls.length).toEqual(1);
  expect(props.onChangePage.mock.calls[0][0]).toEqual({
    page: 2,
    skip: props.limit,
  });
});

test("on click next", () => {
  const { getByTestId } = render(<PaginationButtons {...props} />);

  fireEvent.click(getByTestId("next"));

  expect(props.onChangePage.mock.calls.length).toEqual(1);
  expect(props.onChangePage.mock.calls[0][0]).toEqual({
    page: 4,
    skip: 3 * props.limit,
  });
});

test("on click last", () => {
  const { getByText } = render(<PaginationButtons {...props} />);

  fireEvent.click(getByText("Last"));

  expect(props.onChangePage.mock.calls.length).toEqual(1);
  expect(props.onChangePage.mock.calls[0][0]).toEqual({
    page: 5,
    skip: 4 * props.limit,
  });
});
