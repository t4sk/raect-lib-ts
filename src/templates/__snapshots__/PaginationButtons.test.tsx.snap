// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders 1`] = `
<div>
  <div
    class="component"
  >
    <button
      class="ui mini basic button"
    >
      First
    </button>
    <button
      class="ui mini basic icon button"
      data-testid="prev"
    >
      <i
        aria-hidden="true"
        class="chevron left icon"
      />
    </button>
    <div
      class="ui basic label"
    >
      <div>
        41
        -
        60
         of
         
        100
      </div>
    </div>
    <button
      class="ui mini basic icon button"
      data-testid="next"
    >
      <i
        aria-hidden="true"
        class="chevron right icon"
      />
    </button>
    <button
      class="ui mini basic button"
    >
      Last
    </button>
  </div>
</div>
`;

exports[`render first page 1`] = `
<div>
  <div
    class="component"
  >
    <button
      class="ui mini basic disabled button"
      disabled=""
      tabindex="-1"
    >
      First
    </button>
    <button
      class="ui mini basic icon disabled button"
      data-testid="prev"
      disabled=""
      tabindex="-1"
    >
      <i
        aria-hidden="true"
        class="chevron left icon"
      />
    </button>
    <div
      class="ui basic label"
    >
      <div>
        1
        -
        20
         of
         
        100
      </div>
    </div>
    <button
      class="ui mini basic icon button"
      data-testid="next"
    >
      <i
        aria-hidden="true"
        class="chevron right icon"
      />
    </button>
    <button
      class="ui mini basic button"
    >
      Last
    </button>
  </div>
</div>
`;

exports[`render last page 1`] = `
<div>
  <div
    class="component"
  >
    <button
      class="ui mini basic button"
    >
      First
    </button>
    <button
      class="ui mini basic icon button"
      data-testid="prev"
    >
      <i
        aria-hidden="true"
        class="chevron left icon"
      />
    </button>
    <div
      class="ui basic label"
    >
      <div>
        81
        -
        100
         of
         
        100
      </div>
    </div>
    <button
      class="ui mini basic icon disabled button"
      data-testid="next"
      disabled=""
      tabindex="-1"
    >
      <i
        aria-hidden="true"
        class="chevron right icon"
      />
    </button>
    <button
      class="ui mini basic disabled button"
      disabled=""
      tabindex="-1"
    >
      Last
    </button>
  </div>
</div>
`;
